Add `fcntl_getlk` for fetching information of a lock held by another process

metent opened the pull request:
Fixes #1076.

I added this inside the `process` module as `Pid` was inaccessible in the `fs` module, and it makes more sense to include it inside `process` as we can use it to access the pid, and lock data which is held by another process. I read the code for `rustix::fs::fcntl_lock`, and added this functionality in a similar manner, but let me know if the APIs/code organization need work.

I still need to add a test for this though, but I'm not sure what would be the best way to do that. As a process will always be able to acquire it's own lock, we'll probably need to spawn another process and read the advisory lock set by the parent process over there. We may use `std::os::unix::process::CommandExt::before_exec` for forking but it might not be the best way, given that it's unsafe and we'll have to handle race conditions, so I need some guidance here.

EDIT: Used `std::os::unix::process::CommandExt::pre_exec` for spawning child processes in tests for now.

EDIT 2: After running the tests, I found that MacOS and FreeBSD are the only platform which don't return `EINVAL` when `l_type` is set to `F_UNLCK`. For ensuring consistent behavior, I've added a check [here](https://github.com/bytecodealliance/rustix/pull/1310/files#diff-3f3e90e7733855d35135272bb9f04f01aef49e024cbcb90afe5da63a86ee5b53R669-R671). Let me know if I should respect platform-specific behavior or keep the status quo.

metent commented:
Apparently someone else added a PR for this before(https://github.com/bytecodealliance/rustix/pull/1077). I couldn't find such issue or PR the last time I searched.

sunfishcode commented on src/process/types.rs:
```diff
@@ -0,0 +1,84 @@
+    /// Starting offset for lock
+    pub start: isize,
+    /// Number of bytes to lock
+    pub length: isize,
```
Rustix exposes file sizes and offsets as 64-bit whenever possible, and follows [`std::io::SeekFrom`](https://doc.rust-lang.org/stable/std/io/enum.SeekFrom.html) in using `u64` for absolute offsets. Could you use `u64` for the start and length?

sunfishcode commented on src/backend/libc/process/syscalls.rs:
```diff
@@ -646,3 +655,27 @@ pub(crate) fn getgroups(buf: &mut [Gid]) -> io::Result<usize> {
+pub(crate) fn fcntl_getlk(fd: BorrowedFd<'_>, lock: &Flock) -> io::Result<Option<Flock>> {
+    if lock.typ == crate::process::FlockType::Unlocked {
+        return Err(io::Errno::INVAL);
+    }
```
Is it necessary to do this check here, or would it be ok to just expose the raw behavior of the `fcntl` here? If we have a check like this, we should document it.

metent replied:
I think we should keep the check. Here's my reasoning:

The POSIX specification of `F_GETLK` states that the third argument `arg` is the **lock** description and it explictly mentions the tested creation of the lock:

> Get any lock which blocks the process-owned file **lock description pointed to by the third argument, arg**, taken as a pointer to type struct flock, defined in <fcntl.h>. The information retrieved shall overwrite the information passed to fcntl() in the structure flock. If no lock is found that would prevent **this lock from being created**, then the structure shall be left unchanged except for the lock type in l_type which shall be set to F_UNLCK.

The 'ERRORS' sections states that if the `cmd` argument is `F_GETLK` and the data pointed by it is invalid, then the error is set to `EINVAL`.

> ERRORS
> The fcntl() function shall fail if:
> 
> [EINVAL]
> The cmd argument is invalid; or the cmd argument is F_DUPFD, F_DUPFD_CLOEXEC, or F_DUPFD_CLOFORK and arg is negative or is greater than or equal to {OPEN_MAX}; or the cmd argument is F_SETOWN_EX and the type member of the f_owner_ex structure pointed to by arg is invalid, or the pid member is negative and the type member is F_OWNER_PID or F_OWNER_PGRP; **or the cmd argument is F_GETLK**, F_SETLK, F_SETLKW, F_OFD_GETLK, F_OFD_SETLK, or F_OFD_SETLKW **and the data pointed to by arg is not valid**, or fildes refers to a file that does not support locking.

So, according to POSIX `arg` should be a valid lock description which tests the creation of the lock, and if `l_type` is set to `F_UNLCK` it's a request to unlock, so it should be an invalid argument.

Furthermore, I tried running `fcntl_getlk` with `l_type` set to `F_UNLCK` in MacOS CI, and the result was a no-op (No mutation of passed flock and no error returned). This may also be the case for `FreeBSD` and other OS's, and I think it's best to keep this check for every target, so that we don't have to investigate platform-specific behavior.

I noticed that `FreeBSD` doesn't follow POSIX specification for `fadvise` either, and in `rustix::fd::fadvise`, there is some additional logic for ensuring POSIX compatibility. I'll document the code in a similar way for now, but let me know if something else needs to be done.

sunfishcode replied:
The `fadvise` case is technically about bridging the gap from POSIX's signed `off_t` to Rust's convention of using unsigned `u64`, which rustix mentions in src/lib.rs. That said, looking at it now, I agree that we should clarify that in the documentation. I'll submit a PR.

Looking a `F_GETLK` more closely, calling `fcntl_getlk` with `l_type` set to `F_UNLCK` looks like it would be a bug; application code shouldn't ever want to do that. There are a lot of minor portability hazards and POSIX nonconformances that rustix doesn't hide, so my sense is that we shouldn't make an exception and hide this one. It certainly makes sense to document it though,

sunfishcode commented on src/backend/libc/process/syscalls.rs:
```diff
@@ -646,3 +655,27 @@ pub(crate) fn getgroups(buf: &mut [Gid]) -> io::Result<usize> {
+    let mut curr_lock: c::flock = lock.as_raw();
+    unsafe { ret(c::fcntl(borrowed_fd(fd), c::F_GETLK, &mut curr_lock))? };
+
+    if curr_lock.l_pid == Pid::as_raw(lock.pid) {
```
Could you comment on what this code is doing? I'm looking at the documentation for `F_GETLK` and it's not clear what this condition corresponds to.

sunfishcode commented on src/backend/linux_raw/c.rs:
```diff
@@ -184,12 +184,21 @@ pub(crate) const CLONE_CHILD_SETTID: c_int = linux_raw_sys::general::CLONE_CHILD
 
+#[cfg(feature = "process")]
+#[cfg(target_pointer_width = "32")]
+pub(crate) use linux_raw_sys::general::{flock64 as flock, F_GETLK64 as F_GETLK};
```
Rustix does often alias `*64` things to non`*64` names, but in the case of `F_GETLK64`, it's used with `__NR_fcntl64` which is not currently aliased. Could you either remove this `F_GETLK64` alias and just use `F_GETLK64` in the code, or, if you prefer, add an `__NR_fcntl` alias for `__NR_fcntl64` alias, and use that?

sunfishcode reviewed (commented):
Thanks!

I think putting this in `rustix::process` makes sense, for this and for furture fcntl-style lock functions which expose the full fcntl-style process-associated lock API. The current `rustix::fs::fcntl_getlk` doesn't expose the full fcntl-style lock functionality, and is mainly intended as an alternative for `flock` on platforms which lack `flock`. So we can evolve toward `rustix::process::fcntl_*` eventually being the "full" fcntl-style lock API.

metent commented:
I believe I've addressed everything, let me know if any other change is needed.

sunfishcode reviewed (commented):
This looks great! I just have the one comment above. While I understand the desire to add code to hide platform differences, rustix in general doesn't do that, especially for error conditions, an I'm reluctant to start doing it here.

metent commented:
> There are a lot of minor portability hazards and POSIX nonconformances that rustix doesn't hide

Makes sense if rustix is aiming for a relatively low-level API. I've removed the check and updated the docs and tests accordingly.

sunfishcode commented:
Thanks!

sunfishcode commented:
This is now released in rustix 1.0.0.
